wire-ios/Wire-iOS/Sources/UserInterface/E2EIdentity/E2eINotificationActionsHandler.swift

```diff
@@ -32,22 +33,158 @@ final class E2eINotificationActionsHandler: E2eINotificationActions {
 
     private var enrollCertificateUseCase: EnrollE2eICertificateUseCaseInterface?
     private var snoozeCertificateEnrollmentUseCase: SnoozeCertificateEnrollmentUseCaseProtocol?
+    private let gracePeriodRepository: GracePeriodRepository
+    private let targetVC: UIViewController
 
     // MARK: - Life cycle
 
-    init(enrollCertificateUseCase: EnrollE2eICertificateUseCaseInterface?,
-         snoozeCertificateEnrollmentUseCase: SnoozeCertificateEnrollmentUseCaseProtocol?) {
-        self.enrollCertificateUseCase = enrollCertificateUseCase
-        self.snoozeCertificateEnrollmentUseCase = snoozeCertificateEnrollmentUseCase
+    init(
+        enrollCertificateUseCase: EnrollE2eICertificateUseCaseInterface?,
+        snoozeCertificateEnrollmentUseCase: SnoozeCertificateEnrollmentUseCaseProtocol?,
+        gracePeriodRepository: GracePeriodRepository,
+        targetVC: UIViewController) {
+            self.enrollCertificateUseCase = enrollCertificateUseCase
+            self.snoozeCertificateEnrollmentUseCase = snoozeCertificateEnrollmentUseCase
+            self.gracePeriodRepository = gracePeriodRepository
+            self.targetVC = targetVC
+        }
+
+    public func getCertificate() async {
+        let oauthUseCase = OAuthUseCase(rootViewController: targetVC)
+        do {
+            try await enrollCertificateUseCase?.invoke(authenticate: oauthUseCase.invoke)
+            await confirmSuccessfulEnrollment()
+        } catch {
+            guard let endOfGracePeriod = gracePeriodRepository.fetchEndGracePeriodDate() else {
+                return
+            }
+            await showGetCertificateErrorAlert(canCancel: !endOfGracePeriod.isInThePast)
+        }
     }
 
-    public func enrollCertificate() {
-        // TODO: [WPB-5496] enroll certificate
-        snoozeCertificateEnrollmentUseCase?.stop()
+    public func updateCertificate() async {
+        // TODO: [WPB-3324] update certificate
     }
 
     public func snoozeReminder() async {
-        await snoozeCertificateEnrollmentUseCase?.start()
+        guard let endOfGracePeriod = gracePeriodRepository.fetchEndGracePeriodDate(),
+              endOfGracePeriod.timeIntervalSinceNow > 0,
+              let formattedDuration = durationFormatter.string(from: endOfGracePeriod.timeIntervalSinceNow) else {
+            return
+        }
+
+        let alert = await UIAlertController.reminderGetCertificate(timeLeft: formattedDuration) {
+            Task {
+                await self.snoozeCertificateEnrollmentUseCase?.start()
+            }
+        }
+        await targetVC.present(alert, animated: true)
     }
 
+    // MARK: - Helpers
+
+    private func showGetCertificateErrorAlert(canCancel: Bool) async {
+        let alert = await UIAlertController.getCertificateFailed(canCancel: canCancel) {
```

Comment:
Is it safe to retain self here ?

Reply:
Self will only be retained as long as the alert is displayed, which is fine I guess.